File: app/csharp/AccountingSystem.Domain/Aggregates/JournalEntryAggregate.cs
namespace AccountingSystem.Domain.Aggregates;

using AccountingSystem.Domain.Events;

/// <summary>
/// 仕訳 Aggregate
/// イベントソーシングパターンを適用した仕訳の集約ルート
/// </summary>
public class JournalEntryAggregate
{
    public string Id { get; private set; } = string.Empty;
    public DateOnly EntryDate { get; private set; }
    public string Description { get; private set; } = string.Empty;
    public List<LineItem> LineItems { get; private set; } = new();
    public JournalEntryStatus Status { get; private set; } = JournalEntryStatus.DRAFT;
    public bool Deleted { get; private set; } = false;

    /// <summary>
    /// 未コミットイベント（まだイベントストアに保存されていない）
    /// </summary>
    private readonly List<IEventSourcedDomainEvent> _uncommittedEvents = new();
    public IReadOnlyList<IEventSourcedDomainEvent> UncommittedEvents => _uncommittedEvents.AsReadOnly();

    /// <summary>
    /// バージョン（楽観的ロック用）
    /// </summary>
    public int Version { get; private set; } = 0;

    /// <summary>
    /// イベント再生（Event Replay）
    /// 保存されたイベントから Aggregate の状態を復元
    /// </summary>
    public static JournalEntryAggregate Replay(IReadOnlyList<IEventSourcedDomainEvent> events)
    {
        var aggregate = new JournalEntryAggregate();
        foreach (var @event in events)
        {
            aggregate.Apply(@event);
            aggregate.Version++;
        }
        return aggregate;
    }

    /// <summary>
    /// コマンド: 仕訳を作成
    /// </summary>
    public static JournalEntryAggregate Create(
        string id,
        DateOnly entryDate,
        string description,
        List<LineItem> lineItems,
        string userId)
    {
        // ビジネスルール検証
        if (lineItems == null || lineItems.Count == 0)
        {
            throw new ArgumentException("仕訳明細が必要です");
        }

        var debitTotal = lineItems
            .Where(item => item.DebitCredit == DebitCredit.DEBIT)
            .Sum(item => item.Amount);

        var creditTotal = lineItems
            .Where(item => item.DebitCredit == DebitCredit.CREDIT)
            .Sum(item => item.Amount);

        if (debitTotal != creditTotal)
        {
            throw new ArgumentException(
                $"借方合計と貸方合計が一致しません: 借方={debitTotal}, 貸方={creditTotal}"
            );
        }

        // イベント発行
        var @event = new JournalEntryCreatedEvent
        {
            JournalEntryId = id,
            EntryDate = entryDate,
            Description = description,
            LineItems = lineItems.Select(item =>
                new JournalEntryCreatedEvent.JournalEntryLineItem
                {
                    AccountCode = item.AccountCode,
                    DebitCredit = (JournalEntryCreatedEvent.DebitCreditType)item.DebitCredit,
                    Amount = item.Amount
                }).ToList(),
            UserId = userId,
            OccurredAt = DateTime.UtcNow
        };

        var aggregate = new JournalEntryAggregate();
        aggregate.Apply(@event);
        aggregate._uncommittedEvents.Add(@event);

        return aggregate;
    }

    /// <summary>
    /// コマンド: 仕訳を承認
    /// </summary>
    public void Approve(string approvedBy, string approvalComment)
    {
        if (Deleted)
        {
            throw new InvalidOperationException("削除済みの仕訳は承認できません");
        }

        if (Status == JournalEntryStatus.APPROVED)
        {
            throw new InvalidOperationException("すでに承認済みです");
        }

        var @event = new JournalEntryApprovedEvent
        {
            JournalEntryId = Id,
            ApprovedBy = approvedBy,
            ApprovalComment = approvalComment,
            OccurredAt = DateTime.UtcNow,
            UserId = approvedBy
        };

        Apply(@event);
        _uncommittedEvents.Add(@event);
    }

    /// <summary>
    /// コマンド: 仕訳を削除
    /// </summary>
    public void Delete(string reason, string userId)
    {
        if (Deleted)
        {
            throw new InvalidOperationException("すでに削除済みです");
        }

        var @event = new JournalEntryDeletedEvent
        {
            JournalEntryId = Id,
            Reason = reason,
            OccurredAt = DateTime.UtcNow,
            UserId = userId
        };

        Apply(@event);
        _uncommittedEvents.Add(@event);
    }

    /// <summary>
    /// イベント適用（Apply）
    /// </summary>
    private void Apply(IEventSourcedDomainEvent @event)
    {
        switch (@event)
        {
            case JournalEntryCreatedEvent e:
                Id = e.JournalEntryId;
                EntryDate = e.EntryDate;
                Description = e.Description;
                LineItems = e.LineItems.Select(item => new LineItem(
                    item.AccountCode,
                    (DebitCredit)item.DebitCredit,
                    item.Amount
                )).ToList();
                Status = JournalEntryStatus.DRAFT;
                break;
            case JournalEntryApprovedEvent:
                Status = JournalEntryStatus.APPROVED;
                break;
            case JournalEntryDeletedEvent:
                Deleted = true;
                break;
        }
    }

    /// <summary>
    /// 未コミットイベントをクリア
    /// </summary>
    public void MarkEventsAsCommitted()
    {
        _uncommittedEvents.Clear();
    }

    /// <summary>
    /// 仕訳明細
    /// </summary>
    public class LineItem
    {
        public string AccountCode { get; }
        public DebitCredit DebitCredit { get; }
        public decimal Amount { get; }

        public LineItem(string accountCode, DebitCredit debitCredit, decimal amount)
        {
            AccountCode = accountCode;
            DebitCredit = debitCredit;
            Amount = amount;
        }
    }

    /// <summary>
    /// 借方・貸方区分
    /// </summary>
    public enum DebitCredit
    {
        DEBIT,
        CREDIT
    }

    /// <summary>
    /// 仕訳ステータス
    /// </summary>
    public enum JournalEntryStatus
    {
        DRAFT,
        APPROVED
    }
}

File: app/fsharp/AccountingSystem.Domain/Aggregates/JournalEntryAggregate.fs
namespace AccountingSystem.Domain.Aggregates

open System
open AccountingSystem.Domain.Types
open AccountingSystem.Domain.Events

/// <summary>
/// 仕訳 Aggregate の状態
/// </summary>
type JournalEntryAggregate = {
    Id: string
    EntryDate: DateTime
    Description: string
    LineItems: JournalEntryLineItem list
    Status: JournalEntryStatus
    Deleted: bool
    Version: int
    UncommittedEvents: JournalEntryEvent list
}

module JournalEntryAggregate =
    /// <summary>
    /// 空の Aggregate を作成
    /// </summary>
    let empty : JournalEntryAggregate = {
        Id = ""
        EntryDate = DateTime.MinValue
        Description = ""
        LineItems = []
        Status = JournalEntryStatus.Draft
        Deleted = false
        Version = 0
        UncommittedEvents = []
    }

    /// <summary>
    /// イベント適用（状態遷移）
    /// </summary>
    let applyEvent (aggregate: JournalEntryAggregate) (event: JournalEntryEvent) : JournalEntryAggregate =
        match event with
        | JournalEntryCreated data ->
            { aggregate with
                Id = data.JournalEntryId
                EntryDate = data.EntryDate
                Description = data.Description
                LineItems = data.LineItems
                Status = JournalEntryStatus.Draft
            }

        | JournalEntryApproved _ ->
            { aggregate with Status = JournalEntryStatus.Approved }

        | JournalEntryDeleted _ ->
            { aggregate with Deleted = true }

    /// <summary>
    /// イベント再生（fold で状態を復元）
    /// </summary>
    let replay (events: JournalEntryEvent list) : JournalEntryAggregate =
        events
        |> List.fold (fun aggregate event ->
            { (applyEvent aggregate event) with Version = aggregate.Version + 1 }
        ) empty

    /// <summary>
    /// スナップショットからイベント再生（最適化用）
    /// </summary>
    let replayFrom (snapshot: JournalEntryAggregate) (events: JournalEntryEvent list) : JournalEntryAggregate =
        events
        |> List.fold (fun aggregate event ->
            { (applyEvent aggregate event) with Version = aggregate.Version + 1 }
        ) snapshot

    /// <summary>
    /// ビジネスルール検証: 借方・貸方の合計一致
    /// </summary>
    let validateBalance (lineItems: JournalEntryLineItem list) : Result<unit, string> =
        let debitTotal =
            lineItems
            |> List.filter (fun item -> item.DebitCredit = DebitCreditType.Debit)
            |> List.sumBy (fun item -> item.Amount)

        let creditTotal =
            lineItems
            |> List.filter (fun item -> item.DebitCredit = DebitCreditType.Credit)
            |> List.sumBy (fun item -> item.Amount)

        if debitTotal <> creditTotal then
            Error $"借方合計と貸方合計が一致しません: 借方={debitTotal}, 貸方={creditTotal}"
        else
            Ok ()

    /// <summary>
    /// コマンド: 仕訳を作成
    /// </summary>
    let create
        (id: string)
        (entryDate: DateTime)
        (description: string)
        (lineItems: JournalEntryLineItem list)
        (userId: string)
        : Result<JournalEntryAggregate, string> =

        // ビジネスルール検証
        if List.isEmpty lineItems then
            Error "仕訳明細が必要です"
        else
            match validateBalance lineItems with
            | Error msg -> Error msg
            | Ok () ->
                // イベント発行
                let event = JournalEntryCreated {
                    JournalEntryId = id
                    EntryDate = entryDate
                    Description = description
                    LineItems = lineItems
                    UserId = userId
                    OccurredAt = DateTime.UtcNow
                }

                let aggregate = applyEvent empty event

                Ok { aggregate with
                        Version = 1
                        UncommittedEvents = [event] }

    /// <summary>
    /// コマンド: 仕訳を承認
    /// </summary>
    let approve
        (approvedBy: string)
        (approvalComment: string)
        (aggregate: JournalEntryAggregate)
        : Result<JournalEntryAggregate, string> =

        if aggregate.Deleted then
            Error "削除済みの仕訳は承認できません"
        elif aggregate.Status = JournalEntryStatus.Approved then
            Error "すでに承認済みです"
        else
            let event = JournalEntryApproved {
                JournalEntryId = aggregate.Id
                ApprovedBy = approvedBy
                ApprovalComment = approvalComment
                OccurredAt = DateTime.UtcNow
                UserId = approvedBy
            }

            let updatedAggregate = applyEvent aggregate event

            Ok { updatedAggregate with
                    Version = aggregate.Version + 1
                    UncommittedEvents = aggregate.UncommittedEvents @ [event] }

    /// <summary>
    /// コマンド: 仕訳を削除
    /// </summary>
    let delete
        (reason: string)
        (userId: string)
        (aggregate: JournalEntryAggregate)
        : Result<JournalEntryAggregate, string> =

        if aggregate.Deleted then
            Error "すでに削除済みです"
        else
            let event = JournalEntryDeleted {
                JournalEntryId = aggregate.Id
                Reason = reason
                OccurredAt = DateTime.UtcNow
                UserId = userId
            }

            let updatedAggregate = applyEvent aggregate event

            Ok { updatedAggregate with
                    Version = aggregate.Version + 1
                    UncommittedEvents = aggregate.UncommittedEvents @ [event] }

    /// <summary>
    /// 未コミットイベントをクリア
    /// </summary>
    let markEventsAsCommitted (aggregate: JournalEntryAggregate) : JournalEntryAggregate =
        { aggregate with UncommittedEvents = [] }

    /// <summary>
    /// 借方合計を計算
    /// </summary>
    let sumDebit (aggregate: JournalEntryAggregate) : decimal =
        aggregate.LineItems
        |> List.filter (fun item -> item.DebitCredit = DebitCreditType.Debit)
        |> List.sumBy (fun item -> item.Amount)

    /// <summary>
    /// 貸方合計を計算
    /// </summary>
    let sumCredit (aggregate: JournalEntryAggregate) : decimal =
        aggregate.LineItems
        |> List.filter (fun item -> item.DebitCredit = DebitCreditType.Credit)
        |> List.sumBy (fun item -> item.Amount)
